### PR TITLE
Streaming in-place WACZ creation + CDXJ indexing

### LINKED ISSUE #674: Use in-place streaming to generate WACZ files
The current method of generating WACZ files in the crawler is less than ideal because it:

- Relies on the py-wacz library, despite having removed other Python dependencies in the crawler; and perhaps more importantly
- Requires the crawler to have at least 2x as much storage available as the total size of the WARC files

We can optimize WACZ creation by generating it in-place, without having to copy the WARC file data or read data multiple times.

Details of this approach:

- WACZ contents are streamed to remote upload (or to disk) from existing files on disk
- CDXJ indices are merged using the linux 'sort' command, and compressed to ZipNum if >50K (or always if using --`generateCDX`)

This should result in significant speed / disk usage improvements: previously the WARC was written once, then read again (for CDXJ indexing), read again (for adding to new WACZ ZIP), written to disk (into new WACZ ZIP), and read again (if uploaded to remote endpoint). Now, WARCs are written once along with the CDXJ, CDXJ is merged, and then all data is read once to either generate WACZ on-disk or upload the WACZ to remote storage (i.e. S3 bucket).

This issue supercedes https://github.com/webrecorder/browsertrix-crawler/issues/484, as it offers additional advantages over switching WACZ generation to the js-wacz library.

### PR DESCRIPTION
**Pull request by ikreymer**
Fixes #674 

This PR supersedes #505, and instead of using js-wacz for optimized WACZ creation:
- generates an 'in-place' or 'streaming' WACZ in the crawler, without having to copy the data again.
- WACZ contents are streamed to remote upload (or to disk) from existing files on disk
- CDXJ indices are merged using the linux 'sort' command, and compressed to ZipNum if >50K (or always if using --generateCDX)
- All data in the WACZ is read / written once
- Should result in significant speed / disk usage improvements: previously WARC was written once, then read again (for CDXJ indexing), read again (for adding to new WACZ ZIP), written to disk (into new WACZ ZIP), read again (if upload to remote endpoint). Now, WARCs are written once, along with the per-WARC CDXJ, CDXJ is then merged on-disk, and all data is read once to either generate WACZ on disk or upload to remote.

### CODE REVIEW

File: src/util/wacz.ts
Comment (tw4l) [inline src/util/wacz.ts:311]: One difference from py-wacz I can spot is that the resulting `index.cdx` no longer has a first line of `!meta 0 {"format": "cdxj-gzip-1.0", "filename": "index.cdx.gz"}`, should be simple enough to add

File: tests/basic_crawl.test.js
Comment (tw4l) [inline tests/basic_crawl.test.js:11]: I wonder if it's worth using py-wacz just in some of the tests to validate that the resulting WACZs are valid, or implementing validation in the new `wacz` crawler module to do the same. They do pass validation in my manual testing but would be good to have that in the test suite to catch any potential regressions moving forward given how important this step of the crawling process is.

**Review by tw4l — COMMENTED** [review]
A few initial comments/suggestions, continuing testing though so far it's looking pretty good!

File: src/util/wacz.ts
Comment (tw4l) [reply src/util/wacz.ts:311]: Granted this is possibly unnecessary but seems like a good practice?

File: src/util/wacz.ts
Comment (ikreymer) [reply src/util/wacz.ts:205]: This was adapted from archiveweb.page - where there no actual WARC files on disk, so had to use a fixed date, accounting for timezone, to ensure consistent WACZ files on each download of the WACZ in the extension. However, we do have actual WARC files with last modified dates, so can probably use real dates here.

File: src/util/wacz.ts
Comment (ikreymer) [reply src/util/wacz.ts:204]: As mentioned above, we might not need to do this at all actually.

File: src/util/wacz.ts
Comment (ikreymer) [reply src/util/wacz.ts:307]: ah yeah, forgot about this, should re-add for consistency.

File: src/util/wacz.ts
Comment (ikreymer) [reply src/util/wacz.ts:205]: Edit: removed previous comment, confused myself - we do still need the timezone offset!

File: src/util/wacz.ts
Comment (ikreymer) [reply src/util/wacz.ts:204]: Looking at this again, the timezone if always UTC in the Docker container, unless changed otherwise. But ZIP file contents are generally treated as local timezone, unless an 'extended timestamp field' is used (which I don't think the client-zip library supports) or we pass the offset to the docker container somehow. The result is that when running on non-UTC host, the files in the WACZ do have the wrong timestamp. However, this is already the case with py-wacz, so maybe something we address separately if we want to address it. Otherwise, the WARCs in the WACZ do have a UTC timestamp always.

Comment (tw4l) [reply src/util/wacz.ts]: Interesting. I didn't know that particular quirk of the ZIP format. Seems like maybe we should aim for parity with py-wacz/current behavior for now and create a separate issue to track fixing this on non-UTC hosts.

Comment (ikreymer) [reply src/util/wacz.ts]: yep, just removing the offset stuff altogether keeps that parity, did that for now.

File: tests/basic_crawl.test.js
Comment (ikreymer) [reply tests/basic_crawl.test.js]: yeah, we can run py-wacz outside of Docker and install just for CI, added that

**Comment by ikreymer** [comment]
Updated to latest released warcio.js / wabac.js, should be good to merge when everything else is resolved.

File: src/util/wacz.ts
Comment (ikreymer) [reply src/util/wacz.ts:302]: Added for index.idx in https://github.com/webrecorder/browsertrix-crawler/pull/673/commits/d2c172038bc2b80a28ac9653948a53d8bcc43f51

File: src/util/warcwriter.ts
Comment (ikreymer) [inline src/util/warcwriter.ts:64]: One thing to add regarding tempCdxDir - it's now required to be kept so that we don't have to reindex the WARCs again, if adding to an existing collection. Perhaps should rename to 'warcCdx' or some other name to indicate its not as temporary, eg. it'll be needed again if adding to collection or regenerating WACZ again, so we don't have to reindex.

Comment (ikreymer) [reply src/util/warcwriter.ts:64]: We could regenerate the CDX from WARCs, but no reason to delete them as they're pretty small compared to the WARCs, so just keeping them around in the collection

Comment (tw4l) [reply src/util/warcwriter.ts:64]: Good point! I'm in favor of renaming it, and might be good to document what this is and when you might need it in the crawler docs.

Comment (ikreymer) [reply src/util/warcwriter.ts:64]: Renamed to `warcCdxDir` and using `warc-cdx` instead of `tmp-cdx` for the dir name, in https://github.com/webrecorder/browsertrix-crawler/pull/673/commits/fb427eae146a84e89d8451c037dd5f0ba9464b4f

Comment (tw4l) [reply src/util/warcwriter.ts:64]: That makes sense. Only thing is I think if I didn't have the "dev on the team" perspective, I might be confused why there are indexes in both `warc-cdx` and `indexes` (assuming `--generateCDX` or `--generateWACZ` was used), and how they differ, so would be good to add something to the documentation about this before merging.

File: src/util/wacz.ts
Comment (tw4l) [inline src/util/wacz.ts:331]:
```suggestion
        filename: INDEX_CDX_GZ,
```

Small regression, this should point to the gzipped cdxj, not itself

File: src/util/warcwriter.ts
Comment (tw4l) [reply src/util/warcwriter.ts:64]: Created new issue for that so as not to block this PR: https://github.com/webrecorder/browsertrix-crawler/issues/675

**Review by tw4l — APPROVED** [review]

**Pull request merged**
